### PR TITLE
Roslyn Shim - First Phase

### PR DESCRIPTION
**Pull request by TIHan**
This is the first phase in our Roslyn shim.

We are trying to get rid of IVTs from Roslyn. In order to do this without breaking us, we need to build a shim in Roslyn and consume that shim in VisualFSharp.

### CODE REVIEW

File: RoslynPackageVersion.txt [hunk 1]
```diff
@@ -1 +1 @@
-3.0.0-beta4-19170-01
+3.1.0
```
Comment (TIHan) [inline RoslynPackageVersion.txt]: This will changed.

File: eng/Versions.props [hunk 2]
```diff
@@ -72,6 +72,7 @@
       https://vside.myget.org/F/vs-impl/api/v3/index.json;
       https://myget.org/F/roslyn_concord/api/v3/index.json;
       https://vside.myget.org/F/devcore/api/v3/index.json;
+      C:\roslyn\artifacts\packages\Debug\Release;
```
Comment (TIHan) [inline eng/Versions.props]: This will be removed.

**Comment by auduchinok** [comment]
@TIHan Could you please explain where these `IFSharp*` interfaces are coming from?

**Comment by TIHan** [comment]
They are coming from Roslyn's external access, shown here https://github.com/dotnet/roslyn/pull/34945 .
This is only for VS and has no impact on FCS.

**Comment by TIHan** [comment]
The roslyn shim is ready. When https://github.com/dotnet/roslyn/pull/34945 is merged and has a package, we need to update to that version here.

**Comment by TIHan** [comment]
@auduchinok to answer you specifically, we are trying to get rid of IVTs from Roslyn. In order to do this without breaking us, we need to build a shim in Roslyn and consume that shim in VisualFSharp.

**Comment by forki** [comment]
Can you please expand the description and explain what the shim is doing?

**Comment by sharwell** [comment]
@forki The F# assemblies used several types and members that are defined as `internal` in dotnet/roslyn. The compiler allowed these references to compile because the Roslyn assemblies contained `[InternalsVisibleTo]` attributes specifically allowing F# to reference internal members. Those attributes are going away, so unless the F# assemblies are updated they will no longer compile or run.

As a replacement for the `[InternalsVisibleTo]` attributes, the F# assemblies are being permitted to access the contents of *exactly one* assembly defined in dotnet/roslyn: **Microsoft.CodeAnalysis.ExternalAccess.FSharp**. All APIs defined in other Roslyn assemblies that are needed by F# need to have a definition in this external access assembly, because that is the only way F# assemblies will be able to successfully reference them. These are typically shim APIs, meaning they don't contain substantial logic but instead forward incoming calls from F# to the correct internal Roslyn APIs.

**Comment by forki** [comment]
I see. Thanks for the explanation!

File: src/fsharp/NameResolution.fs [hunk 3]
```diff
@@ -1616,7 +1616,6 @@ type TcResultsSinkImpl(g, ?sourceText: ISourceText) =
             capturedOpenDeclarations.Add(openDeclaration)
 
         member sink.CurrentSourceText = sourceText
-
```
Comment (KevinRansom) [inline src/fsharp/NameResolution.fs]: probably not intended.

File: src/fsharp/fsi/fsi.fsproj [hunk 4]
```diff
@@ -30,6 +30,9 @@
     <ProjectReference Include="..\FSharp.Compiler.Private\FSharp.Compiler.Private.fsproj" />
     <ProjectReference Include="..\FSharp.Compiler.Interactive.Settings\FSharp.Compiler.Interactive.Settings.fsproj" />
 
+    <!-- only used when '$(TargetFramework)' == 'netstandard2.0' -->
```
Comment (KevinRansom) [inline src/fsharp/fsi/fsi.fsproj]: Is this necessary? I don't think fsi references FSharp.Build.dll

**Pull request merged**